contexts/LeveragerContext.tsx

```diff
@@ -0,0 +1,1125 @@
+          const args = [debtManager.address, slippage(limit.borrow)] as const;
+          const gasEstimation = await marketOut.estimateGas.approve(args, opts);
+          hash = await marketOut.write.approve(args, {
+            ...opts,
+            gasLimit: (gasEstimation * GAS_LIMIT_MULTIPLIER) / WEI_PER_ETHER,
+          });
+          break;
+        }
+        default:
+          return;
+      }
+
+      if (!hash) return;
+      await waitForTransaction({ hash });
+    } catch (e: unknown) {
+      setErrorData({ status: true, message: handleOperationError(e) });
+    } finally {
+      setIsLoading(false);
+    }
+  }, [approvalStatus, assetIn, debtManager, limit, marketOut, opts, permit2, userInput]);
+
+  const signPermit = useCallback(
+    async (value: bigint, who: 'assetIn' | 'marketIn' | 'marketOut') => {
+      if (!walletAddress || !maIn || !marketIn || !marketOut || !assetIn || !permit2 || !debtManager) return;
+
+      const deadline = BigInt(dayjs().unix() + 3_600);
+
+      if (who === 'assetIn' && !isPermitAllowed(chain, maIn.assetSymbol)) {
+        const signature = await signTypedDataAsync({
+          primaryType: 'PermitTransferFrom',
+          domain: {
+            name: 'Permit2',
+            chainId: chain.id,
+            verifyingContract: permit2.address,
+          },
+          types: {
+            PermitTransferFrom: [
+              { name: 'permitted', type: 'TokenPermissions' },
+              { name: 'spender', type: 'address' },
+              { name: 'nonce', type: 'uint256' },
+              { name: 'deadline', type: 'uint256' },
+            ],
+            TokenPermissions: [
+              { name: 'token', type: 'address' },
+              { name: 'amount', type: 'uint256' },
+            ],
+          },
+          message: {
+            permitted: {
+              token: assetIn.address,
+              amount: value,
+            },
+            spender: debtManager.address,
+            deadline,
+            nonce: hexToBigInt(
+              keccak256(
+                encodeAbiParameters(
+                  [
+                    { name: 'sender', type: 'address' },
+                    { name: 'token', type: 'address' },
+                    { name: 'assets', type: 'uint256' },
+                    { name: 'deadline', type: 'uint256' },
+                  ],
+                  [walletAddress, assetIn.address, value, deadline],
+                ),
+              ),
+            ),
+          },
+        });
+
+        const permit = {
+          deadline,
+          signature,
+        } as const;
+
+        return { type: 'permit2', value: permit } as const;
+      }
+
+      const [impl, nonce] = await Promise.all([
+        who.startsWith('market')
+          ? publicClient.getStorageAt({
+              address: who === 'marketIn' ? marketIn.address : marketOut.address,
+              slot: '0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc',
```

Comment:
what is this address? can we give it a name?

Reply:
This is the slot address for the implementation address https://eips.ethereum.org/EIPS/eip-1967#logic-contract-address